In [50]:
# imports
import pandas as pd
import seaborn as sns
import numpy as np
import sys
import os

In [51]:
# setting up several paths 
sys.path.append(".")
sys.path.append("..")
sys.path.insert(1, '../scripts')
from dataCleaner import *

In [52]:
# import the data frame
missing_values = ["n/a", "na", "undefined", "?", None]
df = pd.read_csv('../data/Week1_challenge_data_source.csv.bz2', na_values=missing_values)

## basic information

In [53]:
# display the head and tail of the data frame
df

Unnamed: 0     Bearer Id            Start  Start ms              End  \
0                0  1.311450e+19   4/4/2019 12:01     770.0  4/25/2019 14:35   
1                1  1.311450e+19   4/9/2019 13:04     235.0  4/25/2019 08:15   
2                2  1.311450e+19   4/9/2019 17:42       1.0  4/25/2019 11:58   
3                3  1.311450e+19  4/10/2019 00:31     486.0  4/25/2019 07:36   
4                4  1.311450e+19  4/12/2019 20:10     565.0  4/25/2019 10:40   
...            ...           ...              ...       ...              ...   
149996      149996  7.277830e+18  4/29/2019 07:28     451.0  4/30/2019 06:02   
149997      149997  7.349880e+18  4/29/2019 07:28     483.0  4/30/2019 10:41   
149998      149998  1.311450e+19  4/29/2019 07:28     283.0  4/30/2019 10:46   
149999      149999  1.311450e+19  4/29/2019 07:28     696.0  4/30/2019 10:40   
150000      150000           NaN              NaN       NaN              NaN   

        End ms  Dur. (ms)          IMSI  MSISDN/Number          IMEI  ...  \
0        662.0  1823652.0  2.082010e+14   3.366496e+10  3.552120e+13  ...   
1        606.0  1365104.0  2.082020e+14   3.368185e+10  3.579400e+13  ...   
2        652.0  1361762.0  2.082000e+14   3.376063e+10  3.528150e+13  ...   
3        171.0  1321509.0  2.082010e+14   3.375034e+10  3.535660e+13  ...   
4        954.0  1089009.0  2.082010e+14   3.369980e+10  3.540700e+13  ...   
...        ...        ...           ...            ...           ...  ...   
149996   214.0    81230.0  2.082020e+14   3.365069e+10  3.548310e+13  ...   
149997   187.0    97970.0  2.082020e+14   3.366345e+10  3.566050e+13  ...   
149998   810.0    98249.0  2.082020e+14   3.362189e+10  3.572120e+13  ...   
149999   327.0    97910.0  2.082020e+14   3.361962e+10  8.618620e+13  ...   
150000     NaN        NaN           NaN            NaN           NaN  ...   

       Youtube DL (Bytes)  Youtube UL (Bytes)  Netflix DL (Bytes)  \
0              15854611.0          2501332.00          8198936.00   
1              20247395.0         19111729.00         18338413.00   
2              19725661.0         14699576.00         17587794.00   
3              21388122.0         15146643.00         13994646.00   
4              15259380.0         18962873.00         17124581.00   
...                   ...                 ...                 ...   
149996         16191667.0         11763428.00         17883703.00   
149997         13877234.0          8288284.00         19350146.00   
149998         22660510.0          1855903.00          9963942.00   
149999          8817106.0          8305402.00          3322253.00   
150000         11634072.5         11009410.13         11626851.72   

        Netflix UL (Bytes)  Gaming DL (Bytes)  Gaming UL (Bytes)  \
0               9656251.00        278082303.0       1.434415e+07   
1              17227132.00        608750074.0       1.170709e+06   
2               6163408.00        229584621.0       3.956300e+05   
3               1097942.00        799538153.0       1.084972e+07   
4                415218.00        527707248.0       3.529801e+06   
...                    ...                ...                ...   
149996         19678161.00        526609673.0       9.197207e+06   
149997         21293148.00        626893062.0       4.735033e+06   
149998          5065760.00        553539484.0       1.339432e+07   
149999         13172589.00        352536971.0       2.529475e+06   
150000         11001754.82        422044702.6       8.288398e+06   

        Other DL (Bytes)  Other UL (Bytes)  Total UL (Bytes)  Total DL (Bytes)  
0            171744450.0      8.814393e+06        36749741.0       308879636.0  
1            526904238.0      1.505514e+07        53800391.0       653384965.0  
2            410692588.0      4.215763e+06        27883638.0       279807335.0  
3            749039933.0      1.279728e+07        43324218.0       846028530.0  
4            550709500.0      1.391032e+07        38542814.0       5691

In [54]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 56 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Unnamed: 0                                150001 non-null  int64  
 1   Bearer Id                                 149010 non-null  float64
 2   Start                                     150000 non-null  object 
 3   Start ms                                  150000 non-null  float64
 4   End                                       150000 non-null  object 
 5   End ms                                    150000 non-null  float64
 6   Dur. (ms)                                 150000 non-null  float64
 7   IMSI                                      149431 non-null  float64
 8   MSISDN/Number                             148935 non-null  float64
 9   IMEI                                      149429 non-null  float64
 10  Last Location Name  

* unwanted column has been created due to the new compression

In [55]:
# create a data cleaner script instance
cleaner = dataCleaner(df)

Data cleaner in action.


In [56]:
# remove the unwanted column
unwanted_cols = ['Unnamed: 0']
df = cleaner.remove_unwanted_cols(unwanted_cols)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype  
---  ------                                    --------------   -----  
 0   Bearer Id                                 149010 non-null  float64
 1   Start                                     150000 non-null  object 
 2   Start ms                                  150000 non-null  float64
 3   End                                       150000 non-null  object 
 4   End ms                                    150000 non-null  float64
 5   Dur. (ms)                                 150000 non-null  float64
 6   IMSI                                      149431 non-null  float64
 7   MSISDN/Number                             148935 non-null  float64
 8   IMEI                                      149429 non-null  float64
 9   Last Location Name                        148848 non-null  object 
 10  Avg RTT DL (ms)     

* unwanted column removed

In [57]:
# generalized basic numerical information 
df.describe()

Bearer Id       Start ms         End ms     Dur. (ms)          IMSI  \
count  1.490100e+05  150000.000000  150000.000000  1.500000e+05  1.494310e+05   
mean   1.013887e+19     499.188200     498.800880  1.046086e+05  2.082017e+14   
std    2.893170e+18     288.611834     288.097653  8.103762e+04  2.149289e+10   
min    6.917540e+18       0.000000       0.000000  7.142000e+03  2.040470e+14   
25%    7.349880e+18     250.000000     251.000000  5.744050e+04  2.082010e+14   
50%    7.349880e+18     499.000000     500.000000  8.639900e+04  2.082020e+14   
75%    1.304240e+19     749.000000     750.000000  1.324302e+05  2.082020e+14   
max    1.318650e+19     999.000000     999.000000  1.859336e+06  2.140740e+14   

       MSISDN/Number          IMEI  Avg RTT DL (ms)  Avg RTT UL (ms)  \
count   1.489350e+05  1.494290e+05    122172.000000    122189.000000   
mean    4.188282e+10  4.847454e+13       109.795706        17.662883   
std     2.447443e+12  2.241637e+13       619.782739        84.793524   
min     3.360100e+10  4.400150e+11         0.000000         0.000000   
25%     3.365130e+10  3.546070e+13        32.000000         2.000000   
50%     3.366371e+10  3.572200e+13        45.000000         5.000000   
75%     3.368349e+10  8.611970e+13        70.000000        15.000000   
max     8.823970e+14  9.900120e+13     96923.000000      7120.000000   

       Avg Bearer TP DL (kbps)  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
count            150000.000000  ...        1.500010e+05        1.500010e+05   
mean              13300.045927  ...        1.163407e+07        1.100941e+07   
std               23971.878541  ...        6.710569e+06        6.345423e+06   
min                   0.000000  ...        5.300000e+01        1.050000e+02   
25%                  43.000000  ...        5.833501e+06        5.517965e+06   
50%                  63.000000  ...        1.161602e+07        1.101345e+07   
75%               19710.750000  ...        1.744852e+07        1.651556e+07   
max              378160.000000  ...        2.325910e+07        2.201196e+07   

       Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
count        1.500010e+05        1.500010e+05       1.500010e+05   
mean         1.162685e+07        1.100175e+07       4.220447e+08   
std          6.725218e+06        6.359490e+06       2.439675e+08   
min          4.200000e+01        3.500000e+01       2.516000e+03   
25%          5.777156e+06        5.475981e+06       2.104733e+08   
50%          1.164222e+07        1.099638e+07       4.234081e+08   
75%          1.747048e+07        1.650727e+07       6.331742e+08   
max          2.325919e+07        2.201196e+07       8.434419e+08   

       Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  \
count       1.500010e+05      1.500010e+05      1.500010e+05   
mean        8.288398e+06      4.211005e+08      8.264799e+06   
std         4.782700e+06      2.432050e+08      4.769004e+06   
min         5.900000e+01      3.290000e+03      1.480000e+02   
25%         4.128476e+06      2.101869e+08      4.145943e+06   
50%         8.291208e+06      4.218030e+08      8.267071e+06   
75%         1.243162e+07      6.316918e+08      1.238415e+07   
max         1.655879e+07      8.434425e+08      1.655882e+07   

       Total UL (Bytes)  Total DL (Bytes)  
count      1.500000e+05      1.500000e+05  
mean       4.112121e+07      4.546434e+08  
std        1.127639e+07      2.441429e+08  
min        2.866892e+06      7.114041e+06  
25%        3.322201e+07      2.431068e+08  
50%        4.114331e+07      4.558411e+08  
75%        4.903424e+07      6.657055e+08  
max        7.833131e+07      9.029696e+08  

[8 rows x 50 columns]

## data types

In [58]:
# get the string or object data  type features
df.select_dtypes(include=['object'])

Start              End Last Location Name  \
0        4/4/2019 12:01  4/25/2019 14:35        9.16457E+15   
1        4/9/2019 13:04  4/25/2019 08:15            L77566A   
2        4/9/2019 17:42  4/25/2019 11:58            D42335A   
3       4/10/2019 00:31  4/25/2019 07:36            T21824A   
4       4/12/2019 20:10  4/25/2019 10:40            D88865A   
...                 ...              ...                ...   
149996  4/29/2019 07:28  4/30/2019 06:02            D20434A   
149997  4/29/2019 07:28  4/30/2019 10:41            D10223C   
149998  4/29/2019 07:28  4/30/2019 10:46            T51102A   
149999  4/29/2019 07:28  4/30/2019 10:40            L88342B   
150000              NaN              NaN                NaN   

       Handset Manufacturer                 Handset Type  
0                   Samsung   Samsung Galaxy A5 Sm-A520F  
1                   Samsung  Samsung Galaxy J5 (Sm-J530)  
2                   Samsung     Samsung Galaxy A8 (2018)  
3                       NaN                          NaN  
4                   Samsung             Samsung Sm-G390F  
...                     ...                          ...  
149996                Apple  Apple iPhone 8 Plus (A1897)  
149997                Apple      Apple iPhone Se (A1723)  
149998                Apple      Apple iPhone Xs (A2097)  
149999               Huawei               Huawei Fig-Lx1  
150000                  NaN                          NaN  

[150001 rows x 5 columns]

* we have 3 string data type features
* we have 2 datetime data type features

In [59]:
# convert start and end to datetime objects
df = cleaner.convert_to_datetime(df=df)

In [60]:
df[['Start', 'End']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150001 entries, 0 to 150000
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   Start   150000 non-null  datetime64[ns]
 1   End     150000 non-null  datetime64[ns]
dtypes: datetime64[ns](2)
memory usage: 2.3 MB


* successfully converted to datetime objects

In [61]:
# get the numerical data  type features
df.select_dtypes(include=['float64'])

Bearer Id  Start ms  End ms  Dur. (ms)          IMSI  \
0       1.311450e+19     770.0   662.0  1823652.0  2.082010e+14   
1       1.311450e+19     235.0   606.0  1365104.0  2.082020e+14   
2       1.311450e+19       1.0   652.0  1361762.0  2.082000e+14   
3       1.311450e+19     486.0   171.0  1321509.0  2.082010e+14   
4       1.311450e+19     565.0   954.0  1089009.0  2.082010e+14   
...              ...       ...     ...        ...           ...   
149996  7.277830e+18     451.0   214.0    81230.0  2.082020e+14   
149997  7.349880e+18     483.0   187.0    97970.0  2.082020e+14   
149998  1.311450e+19     283.0   810.0    98249.0  2.082020e+14   
149999  1.311450e+19     696.0   327.0    97910.0  2.082020e+14   
150000           NaN       NaN     NaN        NaN           NaN   

        MSISDN/Number          IMEI  Avg RTT DL (ms)  Avg RTT UL (ms)  \
0        3.366496e+10  3.552120e+13             42.0              5.0   
1        3.368185e+10  3.579400e+13             65.0              5.0   
2        3.376063e+10  3.528150e+13              NaN              NaN   
3        3.375034e+10  3.535660e+13              NaN              NaN   
4        3.369980e+10  3.540700e+13              NaN              NaN   
...               ...           ...              ...              ...   
149996   3.365069e+10  3.548310e+13             32.0              0.0   
149997   3.366345e+10  3.566050e+13             27.0              2.0   
149998   3.362189e+10  3.572120e+13             43.0              6.0   
149999   3.361962e+10  8.618620e+13             37.0              5.0   
150000            NaN           NaN              NaN              NaN   

        Avg Bearer TP DL (kbps)  ...  Youtube DL (Bytes)  Youtube UL (Bytes)  \
0                          23.0  ...          15854611.0          2501332.00   
1                          16.0  ...          20247395.0         19111729.00   
2                           6.0  ...          19725661.0         14699576.00   
3                          44.0  ...          21388122.0         15146643.00   
4                           6.0  ...          15259380.0         18962873.00   
...                         ...  ...                 ...                 ...   
149996                     52.0  ...          16191667.0         11763428.00   
149997                     23.0  ...          13877234.0          8288284.00   
149998                     43.0  ...          22660510.0          1855903.00   
149999                     34.0  ...           8817106.0          8305402.00   
150000                      NaN  ...          11634072.5         11009410.13   

        Netflix DL (Bytes)  Netflix UL (Bytes)  Gaming DL (Bytes)  \
0               8198936.00          9656251.00        278082303.0   
1              18338413.00         17227132.00        608750074.0   
2              17587794.00          6163408.00        229584621.0   
3              13994646.00          1097942.00        799538153.0   
4              17124581.00           415218.00        527707248.0   
...                    ...                 ...                ...   
149996         17883703.00         19678161.00        526609673.0   
149997         19350146.00         21293148.00        626893062.0   
149998          9963942.00          5065760.00        553539484.0   
149999          3322253.00         13172589.00        352536971.0   
150000         11626851.72         11001754.82        422044702.6   

        Gaming UL (Bytes)  Other DL (Bytes)  Other UL (Bytes)  \
0            1.434415e+07       171744450.0      8.814393e+06   
1            1.170709e+06       526904238.0      1.505514e+07   
2            3.956300e+05       410692588.0      4.215763e+06   
3            1.084972e+07       749039933.0      1.279728e+07   
4            3.529801e+06       550709500.0      1.391032e+07   
...                   ...               ...               ...   
149996       9.197207e+06         3264510.0      1.348742e+07   
149997       4.7350

* 50 of the columns (features) are numerical data types

In [62]:
# total rows and columns
print(f'This data set  has {df.shape[0]} rows and {df.shape[1]} columns')

This data set  has 150001 rows and 55 columns


## column names

In [63]:
# column names
df.columns.tolist()

['Bearer Id',
 'Start',
 'Start ms',
 'End',
 'End ms',
 'Dur. (ms)',
 'IMSI',
 'MSISDN/Number',
 'IMEI',
 'Last Location Name',
 'Avg RTT DL (ms)',
 'Avg RTT UL (ms)',
 'Avg Bearer TP DL (kbps)',
 'Avg Bearer TP UL (kbps)',
 'TCP DL Retrans. Vol (Bytes)',
 'TCP UL Retrans. Vol (Bytes)',
 'DL TP < 50 Kbps (%)',
 '50 Kbps < DL TP < 250 Kbps (%)',
 '250 Kbps < DL TP < 1 Mbps (%)',
 'DL TP > 1 Mbps (%)',
 'UL TP < 10 Kbps (%)',
 '10 Kbps < UL TP < 50 Kbps (%)',
 '50 Kbps < UL TP < 300 Kbps (%)',
 'UL TP > 300 Kbps (%)',
 'HTTP DL (Bytes)',
 'HTTP UL (Bytes)',
 'Activity Duration DL (ms)',
 'Activity Duration UL (ms)',
 'Dur. (ms).1',
 'Handset Manufacturer',
 'Handset Type',
 'Nb of sec with 125000B < Vol DL',
 'Nb of sec with 1250B < Vol UL < 6250B',
 'Nb of sec with 31250B < Vol DL < 125000B',
 'Nb of sec with 37500B < Vol UL',
 'Nb of sec with 6250B < Vol DL < 31250B',
 'Nb of sec with 6250B < Vol UL < 37500B',
 'Nb of sec with Vol DL < 6250B',
 'Nb of sec with Vol UL < 1250B',
 'Socia

* these are the feature names

## Missing values

In [64]:
df.isna().sum()

Bearer Id                                      991
Start                                            1
Start ms                                         1
End                                              1
End ms                                           1
Dur. (ms)                                        1
IMSI                                           570
MSISDN/Number                                 1066
IMEI                                           572
Last Location Name                            1153
Avg RTT DL (ms)                              27829
Avg RTT UL (ms)                              27812
Avg Bearer TP DL (kbps)                          1
Avg Bearer TP UL (kbps)                          1
TCP DL Retrans. Vol (Bytes)                  88146
TCP UL Retrans. Vol (Bytes)                  96649
DL TP < 50 Kbps (%)                            754
50 Kbps < DL TP < 250 Kbps (%)                 754
250 Kbps < DL TP < 1 Mbps (%)                  754
DL TP > 1 Mbps (%)             

* as you can see there are several missing values

In [65]:
# get percentage missing of the dataframe
cleaner.percent_missing(df)

The dataset contains 12.7195030821 % missing values.


* This is not a small value. 12.72 % is actually missing from the data set.
* Further missing value investigation is required.

* we start from the categorical features, and fill them with their mode

In [66]:
categorical_features = ['Last Location Name', 'Handset Manufacturer', 'Handset Type']
print('missing count')
for feature in categorical_features:
    print(f'{feature} - {df[feature].isnull().sum()}')

missing count
Last Location Name - 1153
Handset Manufacturer - 9559
Handset Type - 9559


In [67]:
loc_mode = df['Last Location Name'].mode()
manufacture_mode = df['Handset Manufacturer'].mode()
type_mode = df['Handset Type'].mode()
print(f'{type(loc_mode.values[0])} Last location mode value: {loc_mode.values[0]}')
print(f'{type(manufacture_mode.values[0])} Handset manufacturer mode value: {manufacture_mode.values[0]}')
print(f'{type(type_mode.values[0])} Handset type mode value: {type_mode.values[0]}')

<class 'str'> Last location mode value: 9.16457E+15
<class 'str'> Handset manufacturer mode value: Apple
<class 'str'> Handset type mode value: Huawei B528S-23A


In [68]:
# fill every categorical feature with the mode
filling_values = {'Last Location Name': loc_mode.values[0], 'Handset Manufacturer': manufacture_mode.values[0], 'Handset Type': type_mode.values[0]}
df.fillna(value=filling_values, inplace=True)

In [69]:
categorical_features = ['Last Location Name', 'Handset Manufacturer', 'Handset Type']
print('missing count')
for feature in categorical_features:
    print(f'{feature} - {df[feature].isnull().sum()}')

missing count
Last Location Name - 0
Handset Manufacturer - 0
Handset Type - 0


* We can see now that all the categorical features has been filled in with their mode

In [70]:
# calculate remaining missing data
# get percentage missing of the dataframe
cleaner.percent_missing(df)

The dataset contains 12.4737956292 % missing values.


* next we will fill the numerical data type features missing values

* if skewed we will fill them with their median / middle values
* if normally distributed we will fill them with their mean / average values

In [71]:
# see the skew
print(df.skew(axis=0))

Bearer Id                                     0.026667
Start ms                                      0.000968
End ms                                       -0.001163
Dur. (ms)                                     3.952609
IMSI                                         40.992855
MSISDN/Number                               332.155855
IMEI                                          1.071470
Avg RTT DL (ms)                              62.907828
Avg RTT UL (ms)                              28.457415
Avg Bearer TP DL (kbps)                       2.589437
Avg Bearer TP UL (kbps)                       4.503413
TCP DL Retrans. Vol (Bytes)                  15.951809
TCP UL Retrans. Vol (Bytes)                  84.113393
DL TP < 50 Kbps (%)                          -2.297803
50 Kbps < DL TP < 250 Kbps (%)                3.271453
250 Kbps < DL TP < 1 Mbps (%)                 4.566158
DL TP > 1 Mbps (%)                            5.370351
UL TP < 10 Kbps (%)                          -8.985016
10 Kbps < 

C:\Users\f0x-tr0t\AppData\Local\Temp\ipykernel_13888\1744353661.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  print(df.skew(axis=0))


* we will take highly skewed features and fill them with  the median value first

In [72]:
skew = df.skew(axis=0)
pos_skew = skew[skew > 0.5].keys()
neg_skew = skew[skew < -0.5].keys()
normal_dist = skew.between(-0.5, 0.5, inclusive='both')
print(f'--- Normally distributed features ---\n{list(normal_dist[normal_dist == True].keys())}\n length: {len(list(normal_dist[normal_dist == True].keys()))}')
print(f'\n--- Positively skewed features ---\n {list(pos_skew)} \n length: {len(pos_skew)}')
print(f'\n--- Negatively skewed features ---\n {list(neg_skew)} \n length: {len(neg_skew)}')

--- Normally distributed features ---
['Bearer Id', 'Start ms', 'End ms', 'Social Media DL (Bytes)', 'Social Media UL (Bytes)', 'Google DL (Bytes)', 'Google UL (Bytes)', 'Email DL (Bytes)', 'Email UL (Bytes)', 'Youtube DL (Bytes)', 'Youtube UL (Bytes)', 'Netflix DL (Bytes)', 'Netflix UL (Bytes)', 'Gaming DL (Bytes)', 'Gaming UL (Bytes)', 'Other DL (Bytes)', 'Other UL (Bytes)', 'Total UL (Bytes)', 'Total DL (Bytes)']
 length: 19

--- Positively skewed features ---
 ['Dur. (ms)', 'IMSI', 'MSISDN/Number', 'IMEI', 'Avg RTT DL (ms)', 'Avg RTT UL (ms)', 'Avg Bearer TP DL (kbps)', 'Avg Bearer TP UL (kbps)', 'TCP DL Retrans. Vol (Bytes)', 'TCP UL Retrans. Vol (Bytes)', '50 Kbps < DL TP < 250 Kbps (%)', '250 Kbps < DL TP < 1 Mbps (%)', 'DL TP > 1 Mbps (%)', '10 Kbps < UL TP < 50 Kbps (%)', '50 Kbps < UL TP < 300 Kbps (%)', 'UL TP > 300 Kbps (%)', 'HTTP DL (Bytes)', 'HTTP UL (Bytes)', 'Activity Duration DL (ms)', 'Activity Duration UL (ms)', 'Dur. (ms).1', 'Nb of sec with 125000B < Vol DL', 'Nb 

C:\Users\f0x-tr0t\AppData\Local\Temp\ipykernel_13888\158291174.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  skew = df.skew(axis=0)


* there are 2 negatively skewed features

* 29 positively skewed features

* and 19 normally distributed features

In [73]:
def fillWithMedian(df: pd.DataFrame, cols: list) -> pd.DataFrame:
    print(f'columns to be filled with median values: {cols}')
    df[cols] = df[cols].fillna(df[cols].median())
    return df

# fill negatively skewed data with median
df = fillWithMedian(df=df, cols=list(neg_skew))

columns to be filled with median values: ['DL TP < 50 Kbps (%)', 'UL TP < 10 Kbps (%)']


In [74]:
# see if there was an error filling with median
df[list(neg_skew)].isna().sum()

DL TP < 50 Kbps (%)    0
UL TP < 10 Kbps (%)    0
dtype: int64

In [75]:
# fill positively skewed data with median
df = fillWithMedian(df=df, cols=list(pos_skew))

columns to be filled with median values: ['Dur. (ms)', 'IMSI', 'MSISDN/Number', 'IMEI', 'Avg RTT DL (ms)', 'Avg RTT UL (ms)', 'Avg Bearer TP DL (kbps)', 'Avg Bearer TP UL (kbps)', 'TCP DL Retrans. Vol (Bytes)', 'TCP UL Retrans. Vol (Bytes)', '50 Kbps < DL TP < 250 Kbps (%)', '250 Kbps < DL TP < 1 Mbps (%)', 'DL TP > 1 Mbps (%)', '10 Kbps < UL TP < 50 Kbps (%)', '50 Kbps < UL TP < 300 Kbps (%)', 'UL TP > 300 Kbps (%)', 'HTTP DL (Bytes)', 'HTTP UL (Bytes)', 'Activity Duration DL (ms)', 'Activity Duration UL (ms)', 'Dur. (ms).1', 'Nb of sec with 125000B < Vol DL', 'Nb of sec with 1250B < Vol UL < 6250B', 'Nb of sec with 31250B < Vol DL < 125000B', 'Nb of sec with 37500B < Vol UL', 'Nb of sec with 6250B < Vol DL < 31250B', 'Nb of sec with 6250B < Vol UL < 37500B', 'Nb of sec with Vol DL < 6250B', 'Nb of sec with Vol UL < 1250B']


In [76]:
# see if there was an error filling with median
df[list(pos_skew)].isna().sum()

Dur. (ms)                                   0
IMSI                                        0
MSISDN/Number                               0
IMEI                                        0
Avg RTT DL (ms)                             0
Avg RTT UL (ms)                             0
Avg Bearer TP DL (kbps)                     0
Avg Bearer TP UL (kbps)                     0
TCP DL Retrans. Vol (Bytes)                 0
TCP UL Retrans. Vol (Bytes)                 0
50 Kbps < DL TP < 250 Kbps (%)              0
250 Kbps < DL TP < 1 Mbps (%)               0
DL TP > 1 Mbps (%)                          0
10 Kbps < UL TP < 50 Kbps (%)               0
50 Kbps < UL TP < 300 Kbps (%)              0
UL TP > 300 Kbps (%)                        0
HTTP DL (Bytes)                             0
HTTP UL (Bytes)                             0
Activity Duration DL (ms)                   0
Activity Duration UL (ms)                   0
Dur. (ms).1                                 0
Nb of sec with 125000B < Vol DL   

* all positively and negatively skewed features has been filled in with their mean

In [77]:
def fillWithMean(df: pd.DataFrame, cols: list) -> pd.DataFrame:
    print(f'columns to be filled with mean values: {cols}')
    df[cols] = df[cols].fillna(df[cols].mean())
    return df

# fill normally distributed missing features with mean
df = fillWithMean(df=df, cols=list(list(normal_dist[normal_dist == True].keys())))

columns to be filled with mean values: ['Bearer Id', 'Start ms', 'End ms', 'Social Media DL (Bytes)', 'Social Media UL (Bytes)', 'Google DL (Bytes)', 'Google UL (Bytes)', 'Email DL (Bytes)', 'Email UL (Bytes)', 'Youtube DL (Bytes)', 'Youtube UL (Bytes)', 'Netflix DL (Bytes)', 'Netflix UL (Bytes)', 'Gaming DL (Bytes)', 'Gaming UL (Bytes)', 'Other DL (Bytes)', 'Other UL (Bytes)', 'Total UL (Bytes)', 'Total DL (Bytes)']


In [78]:
df[list(normal_dist[normal_dist == True].keys())].isna().sum()

Bearer Id                  0
Start ms                   0
End ms                     0
Social Media DL (Bytes)    0
Social Media UL (Bytes)    0
Google DL (Bytes)          0
Google UL (Bytes)          0
Email DL (Bytes)           0
Email UL (Bytes)           0
Youtube DL (Bytes)         0
Youtube UL (Bytes)         0
Netflix DL (Bytes)         0
Netflix UL (Bytes)         0
Gaming DL (Bytes)          0
Gaming UL (Bytes)          0
Other DL (Bytes)           0
Other UL (Bytes)           0
Total UL (Bytes)           0
Total DL (Bytes)           0
dtype: int64

* all normally distributed features missing values filled in with their means

In [90]:
# check percentage missing
cleaner.percent_missing(df)

The dataset contains 0.0 % missing values.


In [85]:
df.drop(df.tail(1).index, inplace=True)

In [89]:
# check percentage missing
cleaner.percent_missing(df)

The dataset contains 0.0 % missing values.


In [91]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 55 columns):
 #   Column                                    Non-Null Count   Dtype         
---  ------                                    --------------   -----         
 0   Bearer Id                                 150000 non-null  float64       
 1   Start                                     150000 non-null  datetime64[ns]
 2   Start ms                                  150000 non-null  float64       
 3   End                                       150000 non-null  datetime64[ns]
 4   End ms                                    150000 non-null  float64       
 5   Dur. (ms)                                 150000 non-null  float64       
 6   IMSI                                      150000 non-null  float64       
 7   MSISDN/Number                             150000 non-null  float64       
 8   IMEI                                      150000 non-null  float64       
 9   Last Location N

* all our missing features are properly filled

In [93]:
# df['IMEI'].hist()

In [94]:
# sns.displot(data = df, x =df['IMEI'])